### ISSUE TITLE
`inputs` or `workflow` is nil when used in sprig expression that has more than 1 parameter (`withParam` loop)

### ISSUE THREAD

**Issue by aaaaahaaaaa** [argoproj/argo-workflows#8224]
## Checklist

 

* [x] Double-checked my configuration.
* [x] Tested using the latest version.
* [x] Used the Emissary executor.

### What happened/what you expected to happen?

With the following context:
* within a `withParam` loop
* in `argument.parameter.value`
* using a `sprig` function that as more than 1 parameter (e.g. `sprig.replace`)
* referencing `workflow.parameters.paramX` or `inputs.parameters.paramX`

Then, the following error is thrown:
```
  Warning  WorkflowFailed   3s    workflow-controller  failed to evaluate expression: cannot fetch parameters from <nil> (1:23)
 |  sprig.replace(inputs.parameters.paramX, 'YYY', 'Replace XXX') 
 | ......................^
```

If the expression uses a sprig function with 1 parameter, then the error doesn't occur.

### What version are you running?
`3.3.0`

## Diagnostics

Paste the smallest workflow that reproduces the bug. We must be able to run the workflow.

```yaml
---
apiVersion: argoproj.io/v1alpha1
kind: WorkflowTemplate
metadata:
  name: my-template
  namespace: argo
spec:
  entrypoint: execute

  arguments:
    parameters:
      - name: paramX
        value: XXX
      - name: paramY
        valueFrom:
          configMapKeyRef:
            name: my-config
            key: my-key

  templates:
    - name: execute
      steps:
        - - name: other-template
            templateRef:
              name: other-template
              template: execute
            withParam: "{{workflow.parameters.paramY}}"
            arguments:
              parameters:
                - name: my-param
                  value: "{{= sprig.replace(workflow.parameters.paramX, "YYY", "Replace XXX" }}"
```

All the following are however working and are producing the expected values:
```yaml
              parameters:
                - name: my-param
                  value: "{{ workflow.parameters.paramX }}"
```
```yaml
              parameters:
                - name: my-param
                  value: "{{= workflow.parameters.paramX }}"
```
```yaml
              parameters:
                - name: my-param
                  value: "{{= sprig.trim(workflow.parameters.paramX) }}"
```

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

**Comment by aaaaahaaaaa** [comment]
Could be related to https://github.com/argoproj/argo-workflows/issues/6790?

**Comment by sarabala1979** [comment]
@aaaaahaaaaa Is it only happening when you are using `input.parameters.xxxx`?

**Comment by aaaaahaaaaa** [comment]
@sarabala1979 Or `workflow.parameters` as mentioned above.

**Comment by aaaaahaaaaa** [comment]
@sarabala1979 I did some more testing and I think I drilled down to a more specific diagnostic that my example above doesn't show.

The problem seems to occur when both `worflow.parameters.XXX` and `item.YYY` are used next to each other as parameters of the same function.

Examples (as part of a `withParam` loop):

WORKS
```
{{= sprig.replace('what ever', workflow.parameters.XXX, workflow.parameters.XXX) }}
```

WORKS
```
{{= sprig.replace('what ever', item.YYY, item.YYY) }}
```


ERROR
```
{{= sprig.replace('what ever', workflow.parameters.XXX, item.YYY) }}
```

This seems really strange but I believe there's really a bug here.

**Comment by tacf** [comment]
Hi, bumped into an issue that i think it may relate to this. 

There's a weird behaviour when using both `workflow.` and `inputs.` in an expression. In my case i'm trying to dynamically set a mutex, below i'll leave the examples and the calculated mutex value to see if it helps trying to pin point this.


```
workflow.parameters.XXX = true
inputs.parameters.YYY = false

"{{=workflow.parameters.XXX}}" # argo/Mutex/true
"{{=workflow.parameters.XXX == 'true' ? workflow.uid : workflow.parameters.XXX }}" # argo/Mutex/d003efcc-7817-4dc0-b435-5c98640ac358
"{{=inputs.parameters.YYY == 'false' ? workflow.uid : 'falseee' }}" # argo/Mutex/{{=inputs.parameters.YYY == 'false' ? workflow.uid : 'falseee' }}
"{{inputs.parameters.YYY}}" # argo/Mutex/false
"{{=inputs.parameters.YYY}}" # argo/Mutex/false
"{{=workflow.parameters.XXX == 'true' and inputs.parameters.YYY == 'false' ? 'true' : 'false'}}" # argo/Mutex/{{=workflow.parameters.XXX == 'true' and inputs.parameters.YYY == 'false' ? 'true' : 'false'}}
"{{=inputs.parameters.YYY == 'false' and workflow.parameters.XXX == 'true' ? 'true' : 'false'}}" # argo/Mutex/{{=inputs.parameters.YYY == 'false' and workflow.parameters.XXX == 'true' ? 'true' : 'false'}}
"{{=inputs.parameters.YYY == 'false' and inputs.parameters.YYY == 'true' ? 'true' : 'false'}}" # argo/Mutex/false
"{{=workflow.parameters.XXX == 'true' and workflow.parameters.XXX == 'false' ? 'true' : 'false'}}" # argo/Mutex/false 

```

As you can see in the case that both `workflow.` and `inputs.` is used the mutex value goes as the literal string of the expression, and according to the tests i did is not just a visual thing (it could be either an UI / Logs issue no rendering the calculated value) the workflow is actually locking on the expression string.

Let me know if you find this is related or should i move this to a new issue.

**Comment by alexec** [comment]
```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: my-wf-
spec:
  entrypoint: main

  arguments:
    parameters:
      - name: XXX
        value: "true"
      - name: YYY
        value: "false"

  templates:
    - name: main
      inputs:
        parameters:
          - name: XXX
            value: "{{workflow.parameters.XXX}}"
          - name: YYY
            value: "{{workflow.parameters.YYY}}"
      container:
        image: argoproj/argosay:v2
        command:
          - sh
          - -c
        args:
          - |
            echo "{{=workflow.parameters.XXX}}" # argo/Mutex/true
            echo "{{=workflow.parameters.XXX == 'true' ? workflow.uid : workflow.parameters.XXX }}" # argo/Mutex/d003efcc-7817-4dc0-b435-5c98640ac358
            echo "{{=inputs.parameters.YYY == 'false' ? workflow.uid : 'falseee' }}" # argo/Mutex/{{=inputs.parameters.YYY == 'false' ? workflow.uid : 'falseee' }}
            echo "{{inputs.parameters.YYY}}" # argo/Mutex/false
            echo "{{=inputs.parameters.YYY}}" # argo/Mutex/false
            echo "{{=workflow.parameters.XXX == 'true' and inputs.parameters.YYY == 'false' ? 'true' : 'false'}}" # argo/Mutex/{{=workflow.parameters.XXX == 'true' and inputs.parameters.YYY == 'false' ? 'true' : 'false'}}
            echo "{{=inputs.parameters.YYY == 'false' and workflow.parameters.XXX == 'true' ? 'true' : 'false'}}" # argo/Mutex/{{=inputs.parameters.YYY == 'false' and workflow.parameters.XXX == 'true' ? 'true' : 'false'}}
            echo "{{=inputs.parameters.YYY == 'false' and inputs.parameters.YYY == 'true' ? 'true' : 'false'}}" # argo/Mutex/false
            echo "{{=workflow.parameters.XXX == 'true' and workflow.parameters.XXX == 'false' ? 'true' : 'false'}}" # argo/Mutex/false 
```

````
✗ argo submit --log tmp.yaml 
Name:                my-wf-h4p5d
Namespace:           argo
ServiceAccount:      unset (will run with the default ServiceAccount)
Status:              Pending
Created:             Mon Apr 04 07:36:02 -0700 (now)
Progress:            
Parameters:          
  XXX:               true
  YYY:               false

This workflow does not have security context set. You can run your workflow pods more securely by setting it.
Learn more at https://argoproj.github.io/argo-workflows/workflow-pod-security-context/
my-wf-h4p5d: time="2022-04-04T14:36:04.207Z" level=info msg="capturing logs" argo=true
my-wf-h4p5d: true
my-wf-h4p5d: c9f7ac42-fa9d-44fe-8694-7479965a3a32
my-wf-h4p5d: c9f7ac42-fa9d-44fe-8694-7479965a3a32
my-wf-h4p5d: false
my-wf-h4p5d: false
my-wf-h4p5d: true
my-wf-h4p5d: true
my-wf-h4p5d: false
my-wf-h4p5d: false
```

**Comment by aaaaahaaaaa** [comment]
@alexec What about the original issue?

**Comment by alexec** [comment]
Could you please re-state the problem?

**Comment by aaaaahaaaaa** [comment]
@alexec You mean the original post is unclear?

**Comment by aaaaahaaaaa** [comment]
The gist is, considering the following workflow:
```
---
apiVersion: argoproj.io/v1alpha1
kind: WorkflowTemplate
metadata:
  name: my-template
  namespace: argo
spec:
  entrypoint: execute

  arguments:
    parameters:
      - name: X
        value: xxx
      - name: Y
        valueFrom:
          configMapKeyRef:
            name: my-config
            key: my-key

  templates:
    - name: execute
      steps:
        - - name: other-template
            templateRef:
              name: other-template
              template: execute
            withParam: "{{workflow.parameters.Y}}"
            arguments:
              parameters:
                - name: my-param
                  value: "{{= sprig.replace('what ever', workflow.parameters.X, item.Y) }}"
```

Then the following error is raised:
```
  Warning  WorkflowFailed   3s    workflow-controller  failed to evaluate expression: cannot fetch parameters from <nil> (1:23)
 |  sprig.replace(, 'what ever', inputs.parameters.paramX, item.Y) 
 | ............................................^
```

The problem seems to specifically occur when both worflow.parameters.X and item.Y are used next to each other as parameters of the same function. I believe there's clearly a bug here.

**Comment by aaaaahaaaaa** [comment]
@alexec Would you confirm that looks like a bug to you?

**Comment by alexec** [comment]
> It's not a problem with **sprig**. It's a problem with Argo variables expression in general

Could you please explain more?

**Comment by rocketspacer** [comment]
@alexec 

This should work, but it doesn't
```
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: debug-
spec:
  arguments:
    parameters:
      - name: message
        value: "hello"

  entrypoint: entrypoint

  templates:
    - name: entrypoint
      steps:
          - name: broken
            template: speak
            arguments:
              parameters:
                - name: message
                  value: "{{= workflow.parameters.message + ' ' + item }}"
            withSequence:
              count: "3"

    - name: speak
      inputs:
        parameters:
          - name: message
      script:
        image: python:alpine
        command: ["python"]
        source: |-
          print("{{ message }}")
```

## Expectations:
`{{= workflow.parameters.message + ' ' + item }}` should evaluated to `hello 0` `hello 1` `hello 2` for each fanned-out instance of the step **broken**

## Reality
I got
<img width="487" alt="image" src="https://user-images.githubusercontent.com/17026876/165352629-ac94dac8-7a94-44ef-a033-e5b6ec8beb23.png">

**Comment by gg314** [comment]
This might be a long shot, but @aaaaahaaaaa and @rocketspacer, could you try using `workflow.parameters['message']` instead of `workflow.parameters.message`?

**Comment by aaaaahaaaaa** [comment]
> This might be a long shot, but @aaaaahaaaaa and @rocketspacer, could you try using `workflow.parameters['message']` instead of `workflow.parameters.message`?

Already tried. Same outcome.

**Comment by gg314** [comment]
That's a shame. I thought I was having the same problem but switching to this syntax fixed it. There is definitely something strange about the parser...

**Comment by zjgemi** [comment]
I think the problem is that the workflow contexts (workflow.xxx, inputs.xxx, steps.xxx or tasks.xxx) and the loop context (item) cannot be introduced simultaneously for a expression (starts with {{=).

**Comment by ChillAndImprove** [comment]
Up

**Comment by tooptoop4** [comment]
Up

**Comment by philippeboyd** [comment]
@alexec any news on this? I can replicate the issue with a ternary operator

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: debug-
spec:
  arguments:
    parameters:
      - name: default
        value: "HELLO"
      - name: message
        value: |
          [
            {"a" :1},
            {"a" :1,"b": 2}
          ]

  entrypoint: entrypoint

  templates:
    - name: entrypoint
      steps:
        - - name: broken
            template: speak
            arguments:
              parameters:
                - name: message
                  value: "{{=item.b != nil ? item.b : workflow.parameters.default}}"
            withParam: "{{workflow.parameters.message}}"

    - name: speak
      inputs:
        parameters:
          - name: message
      script:
        image: python:alpine
        command: ["python"]
        source: |-
          print("{{ inputs.parameters.message }}")
```

However, replacing the value with `value: "{{=item.b != nil ? item.b : 'BYE'}}"` works. Why can't we reference any `workflow.parameters`?

**Comment by kaaquist** [comment]
up

**Comment by roelarents** [comment]
+1. This makes it impossible to select a field in a (json) parameter by loop item. E.g.: `{{=sprig.fromJson(input.parameters.somejsonparam)[item]}}`

**Comment by Arcahub** [comment]
@alexec Any thoughts about this issue from the maintainers team ? One year I am still facing this issue and I don't understand why this issue is marked as closed.

**Comment by hadams95** [comment]
Would my issue be related to this?

I was attempting to map two input parameters to a single output using Helm's `dig` function.

I have `values.yaml` for instance:
```yaml
map:
  x:
    y:  "z"
```

In my workflowtemplate resource, I have an output parameter:
```yaml
outputs:
  parameters:
  - name: mapped_value # name of output parameter
    value: "{{dig `{{inputs.parameters.x}}` `{{inputs.parameters.y}}` `<should-be-z>` .Values.map}}"
```

If I hardcode the parameters passed to the `dig` function, I get the expected result `"z"`.
**Works**
```yaml
"{{dig x y `<should-be-z>` .Values.map}}"
returns
z
```

However, when I pass the values to the `dig` function via the input parameters, it always gets the fallback value `"<should-be-z>"`.
**Doesn't work**
```yaml
"{{dig `{{inputs.parameters.x}}` `{{inputs.parameters.y}}` `<should-be-z>` .Values.map}}"
returns
<should-be-z>
```

Even when I verify the input parameter values are exactly as I expect them, I cannot get this helm function to work with them.  I can't figure out what I'm doing wrong here.

**Comment by Spenhouet** [comment]
We are also effected by this.

**Comment by edwardfung123** [comment]
looks like I am facing the same problem with a workflow using `withParam`. just part of the code.

```
arguments:
  parameters:
    - name: var-name
      value: "{{=item.a + inputs.parameters.b}}"
```

**Comment by deynekas** [comment]
I am facing same problem
this expression does not work 
`when: '{{=sprig.has(item.batch, sprig.mustFromJson(steps.batches.outputs.result)) }}'`
this does work 
`when: '{{=sprig.has(item.batch, ["sbox","prod"]) }}'`
this also works
`when: '{{=sprig.has("sbox", sprig.mustFromJson(steps.batches.outputs.result)) }}'`

**Comment by thevilledev** [comment]
I ran a few tests on a local kind cluster from a `main` branch build. These work:

- `workflow.parameters.message + '-' + string(item)`
- `workflow.parameters.prefix + '-' + item.name`
- `sprig.replace(workflow.parameters.target, item.greeting, 'say X to the world')`
- `int(item) + int(inputs.parameters.offset)`
- `item.active == 'yes' ? item.name : workflow.parameters.fallback`

And from the previous comment, this works too:

```
when: "{{=sprig.has(item.batch, sprig.mustFromJson(steps.batches.outputs.result)) }}"
```

I think the issue was fixed through #14718, #15033 and #15442. Is this issue ready to be closed?
